From 4e07c396442a758af2b79932e97d655b3c990f83 Mon Sep 17 00:00:00 2001
From: So1aric <51471119+So1aric@users.noreply.github.com>
Date: Thu, 13 Nov 2025 03:21:28 +0800
Subject: [PATCH 055/133] Sema: fix inline fn compiler crash (#25586)
X-Developer-Signature: v=1; a=openpgp-sha256; l=1631; i=xtex@astrafall.org;
 h=from:subject; bh=Y8rxAm/GzhAEXrxMX0+qoF72ba0jw4jRiR7BuPkXRAk=;
 b=owGbwMvMwCW2U4Ij7wZL9ETG02pJDJn78hj/cIjb7Mnf1c/2krv3a8pPMe2t759oeDuIlCsFu
 kbvufKgYyILgxgXg6WYIkuRYYM3q046v+iyclmYOaxMIEOkRRoYGBgYWBj4chPzSo10jPRMtQ31
 DA11jHSMGLg4BWCqby1kZLjD3aVibbiBUzSgsibW91XQip+FHi+aT/04e0GT8a3rtlxGhqmJ+7c
 WnNbqki2xucPx9eFx9sXvNmZyb39zlMN9ZmvGI0YA
X-Developer-Key: i=xtex@astrafall.org; a=openpgp;
 fpr=7231804B052C670F15A6771DB918086ED8045B91

Resolves: https://github.com/ziglang/zig/issues/25581
(cherry picked from commit 49e19fc94fe70cecc1cba21a04c1f2f1c1e4deb8)
---
 src/Sema.zig                                          |  4 +---
 .../compile_errors/value_as_inline_fn_return_type.zig | 11 +++++++++++
 2 files changed, 12 insertions(+), 3 deletions(-)
 create mode 100644 test/cases/compile_errors/value_as_inline_fn_return_type.zig

diff --git a/src/Sema.zig b/src/Sema.zig
index d414c2a32eba..eef065bc1432 100644
--- a/src/Sema.zig
+++ b/src/Sema.zig
@@ -25461,9 +25461,7 @@ fn zirFuncFancy(sema: *Sema, block: *Block, inst: Zir.Inst.Index) CompileError!A
         extra_index += 1;
         if (extra.data.bits.ret_ty_is_generic) break :blk .generic_poison;
 
-        const ret_ty_air_ref = try sema.resolveInst(ret_ty_ref);
-        const ret_ty_val = try sema.resolveConstDefinedValue(block, ret_src, ret_ty_air_ref, .{ .simple = .function_ret_ty });
-        break :blk ret_ty_val.toType();
+        break :blk try sema.resolveType(block, ret_src, ret_ty_ref);
     } else .void;
 
     const noalias_bits: u32 = if (extra.data.bits.has_any_noalias) blk: {
diff --git a/test/cases/compile_errors/value_as_inline_fn_return_type.zig b/test/cases/compile_errors/value_as_inline_fn_return_type.zig
new file mode 100644
index 000000000000..d0d35defa69c
--- /dev/null
+++ b/test/cases/compile_errors/value_as_inline_fn_return_type.zig
@@ -0,0 +1,11 @@
+inline fn a() null {
+    return null;
+}
+
+pub fn main() void {
+    _ = a();
+}
+
+// error
+//
+// :1:15: error: expected type 'type', found '@TypeOf(null)'
-- 
2.52.0

